File: Editor/src/main/java/org/turbomedia/turboedit/editor/panes/preferences/EditRenderServerPane.java
package org.turbomedia.turboedit.editor.panes.preferences;

import atlantafx.base.controls.Spacer;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Cursor;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import org.turbomedia.turboedit.editor.misc.PreferencesFile;
import org.turbomedia.turboedit.editor.renderer.FileResolveMethod;
import org.turbomedia.turboedit.editor.renderer.RenderServerEntry;
import org.turbomedia.turboedit.editor.windows.SimpleMessageBox;

import java.io.IOException;

import static org.turbomedia.turboedit.editor.misc.Locale.GetText;

public class EditRenderServerPane extends VBox {
    public EditRenderServerPane(Stage stage, RenderServerEntry entry) {
        setPadding(new Insets(15));
        setSpacing(6);

        var alreadyExists = PreferencesFile.CURRENT_PREFERENCES.renderServers.contains(entry);

        var serverNameLabel = new Label(GetText("edit_render_server.display_name"));
        var serverNameInput = new TextField(entry.displayName());
        serverNameLabel.setLabelFor(serverNameInput);
        serverNameInput.setDisable(entry.buildIn());

        var serverIPLabel = new Label(GetText("edit_render_server.ip"));
        var serverIPInput = new TextField(entry.ip());
        serverIPLabel.setLabelFor(serverIPInput);
        serverIPInput.setDisable(entry.buildIn());

        var fileModeLabel = new Label(GetText("edit_render_server.file_resolve"));
        var fileModeBox = new ComboBox<FileResolveMethod>();

        for (var item : FileResolveMethod.values()) {
            fileModeBox.getItems().add(item);
        }

        fileModeBox.setDisable(entry.buildIn());
        fileModeBox.setValue(entry.fileResolveMethod());
        fileModeBox.setMaxWidth(Double.MAX_VALUE);
        fileModeBox.setCursor(Cursor.HAND);
        fileModeLabel.setLabelFor(fileModeBox);

        var defaultCheckbox = new CheckBox(GetText("edit_render_server.default_server"));
        defaultCheckbox.setCursor(Cursor.HAND);
        defaultCheckbox.setSelected(entry.defaultServer());

        var addButton = new Button(GetText("edit_render_server.button." + (alreadyExists ? "edit" : "add")));
        addButton.setCursor(Cursor.HAND);
        addButton.setOnAction((event) -> {
            if (serverNameInput.getText().isBlank()) {
                new SimpleMessageBox(GetText("edit_render_server.error.no_display_name.title"), GetText("edit_render_server.error.no_display_name"));
                return;
            }

            var newEntry = new RenderServerEntry(serverNameInput.getText(), serverIPInput.getText(), fileModeBox.getValue(), defaultCheckbox.isSelected(), false);

            try {
                if (alreadyExists) {
                    PreferencesFile.CURRENT_PREFERENCES.removeRenderServers(entry);
                }

                PreferencesFile.CURRENT_PREFERENCES.addRenderServers(newEntry);
            } catch (IOException | InterruptedException e) {
                throw new RuntimeException(e);
            }

            stage.close();
        });

        var closeButton = new Button(GetText("edit_render_server.button.close"));
        closeButton.setCursor(Cursor.HAND);
        closeButton.setOnAction((event) -> stage.close());

        var actionBox = new HBox(defaultCheckbox, new Spacer(), addButton, closeButton);
        actionBox.setAlignment(Pos.CENTER);
        actionBox.setSpacing(6);

        getChildren().addAll(serverNameLabel, serverNameInput, serverIPLabel, serverIPInput, fileModeLabel, fileModeBox, actionBox);
    }
}

File: Editor/src/main/kotlin/turboedit/editor/panes/preferences/EditRenderServerPane.kt
package turboedit.editor.panes.preferences

import atlantafx.base.controls.Spacer
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Cursor
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.Stage
import turboedit.editor.events.EventSystem.callEvent
import turboedit.editor.events.EventType
import turboedit.editor.misc.Locale
import turboedit.editor.misc.PreferencesFile
import turboedit.editor.renderer.FileResolveMethod
import turboedit.editor.renderer.RenderServerEntry
import turboedit.editor.windows.SimpleMessageBox
import java.io.IOException
import java.util.*

class EditRenderServerPane(stage: Stage, entry: RenderServerEntry, responseID: String?) : VBox() {
    init {
        padding = Insets(15.0)
        spacing = 6.0

        val alreadyExists = PreferencesFile.CURRENT_PREFERENCES!!.renderServers.contains(entry)

        val serverNameLabel = Label(Locale.getText("edit_render_server.display_name"))
        val serverNameInput = TextField(entry.displayName())
        serverNameLabel.labelFor = serverNameInput
        serverNameInput.isDisable = entry.buildIn()

        val serverIPLabel = Label(Locale.getText("edit_render_server.ip"))
        val serverIPInput = TextField(entry.ip())
        serverIPLabel.labelFor = serverIPInput
        serverIPInput.isDisable = entry.buildIn()

        val fileModeLabel = Label(Locale.getText("edit_render_server.file_resolve"))
        val fileModeBox = ComboBox<FileResolveMethod?>()

        for (item in FileResolveMethod.entries) {
            fileModeBox.getItems().add(item)
        }

        fileModeBox.isDisable = entry.buildIn()
        fileModeBox.value = entry.fileResolveMethod()
        fileModeBox.maxWidth = Double.Companion.MAX_VALUE
        fileModeBox.cursor = Cursor.HAND
        fileModeLabel.labelFor = fileModeBox

        val defaultCheckbox = CheckBox(Locale.getText("edit_render_server.default_server"))
        defaultCheckbox.cursor = Cursor.HAND
        defaultCheckbox.isSelected = entry.defaultServer()

        val addButton = Button(Locale.getText("edit_render_server.button." + (if (alreadyExists) "edit" else "add")))
        addButton.cursor = Cursor.HAND
        addButton.onAction = EventHandler { event: ActionEvent? ->
            if (serverNameInput.text.isBlank()) {
                SimpleMessageBox(
                    Locale.getText("edit_render_server.error.no_display_name.title"),
                    Locale.getText("edit_render_server.error.no_display_name")
                )

                return@EventHandler
            }
            val newEntry = RenderServerEntry(
                UUID.randomUUID().toString(),
                serverNameInput.text,
                serverIPInput.text,
                fileModeBox.getValue(),
                defaultCheckbox.isSelected,
                entry.buildIn()
            )

            try {
                if (alreadyExists) {
                    PreferencesFile.CURRENT_PREFERENCES!!.removeRenderServers(entry)
                }

                PreferencesFile.CURRENT_PREFERENCES!!.addRenderServers(newEntry)
            } catch (e: IOException) {
                throw RuntimeException(e)
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }

            callEvent(EventType.PREFERENCES_CHANGED, responseID)
            stage.close()
        }

        val closeButton = Button(Locale.getText("edit_render_server.button.close"))
        closeButton.cursor = Cursor.HAND
        closeButton.onAction = EventHandler { event: ActionEvent? -> stage.close() }

        val actionBox = HBox(defaultCheckbox, Spacer(), addButton, closeButton)
        actionBox.alignment = Pos.CENTER
        actionBox.spacing = 6.0

        children.addAll(
            serverNameLabel,
            serverNameInput,
            serverIPLabel,
            serverIPInput,
            fileModeLabel,
            fileModeBox,
            actionBox
        )
    }
}
